core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala

```diff
@@ -585,17 +592,15 @@ class ExternalAppendOnlyMap[K, V, C](
       } else {
         logInfo(s"Task ${context.taskAttemptId} force spilling in-memory map to disk and " +
           s"it will release ${org.apache.spark.util.Utils.bytesToString(getUsed())} memory")
-        nextUpstream = spillMemoryIteratorToDisk(upstream)
+        val nextUpstream = spillMemoryIteratorToDisk(upstream)
+        assert(!upstream.hasNext)
         hasSpilled = true
+        upstream = nextUpstream
```

Comment:
Does the change means we should reassign `upstream` (which eliminates reference to `currentMap`) after  spill **immediately**, otherwise,  we may hit OOM (e.g. never `readNext()` after spill - is this the real cause for JIRA issue?) ?

Reply:
Basically yes, according to my understanding of the code this should have happened on the subsequent hasNext/next call. However according to the analysis in the jira the iterator kept holding this reference, my guess: at this point the entire program started suffering lengthy GC pauses that got it into behaving as if under a deadlock,effectively leaving the ref in place (just a guess)

Reply:
Thanks for fixing this issues. I think the potential solution is to change the `upstream` reference, but I have not tested if this change is sufficient and safe.

Reply:
@jerrylead, I'd appreciate if you could test this.
One more thing that bugs me is that there's another case when the iterator no longer needs the upstream iterator/underlying map but still holds a reference to it:
When the iterator reach EOF its hasNext method returns false which causes the wrapping CompletionIterator to call the cleanup function which simply nulls the underlying map member variable in ExternalAppendOnlyMap, the iterator member is not nulled out so we end up with the CompletionIterator holding two paths to the upstrean iterator which leads to the underlying map: first it still holds a reference to the iterator itself, however it still holds a reference to the cleanup closure which refers the ExternalAppendOnlyMap which still refers to the current iterator which refers upstream...
This can be solven in one of two ways:
Simple way, when creating the completion iterator, provide a closure referring the iterator, not the ExternalAppendOnlyMap.
Thorough way, modify completion iterator to null out references after cleaning up.

Having that said, I'm not sure how long a completed iterator may be 'sitting' before being discarded so I'm not sure if this is worth fixing, especially using the thorough approach.

Reply:
`destroy` re-assigns `upstream`, once `destroy` is called, we should be fine?

Reply:
@cloud-fan , the assumption here is that there are two references to the underlying map: the upstream iterator and the external map itself.
the destroy method first removes the ref via upstream-iterator than delegates to the method that clears the ref via the external map member (currentMap I think), so unless we've missed another ref we should be fine.

as I wrote above, I think there's a potentially more fundamental issue with `CompletionIterator` which keeps holding references via it's `sub` and `completionFunction` members , these might stall some objects from being collected and can be eliminated upon exhaustion of the iterator. there might be some more 'candidates' like `LazyIterator` and `InterruptibleIterator`, I think this desrves some more investigation.

Reply:
@cloud-fan , do you think this is worth doing, I'm referring to the CompletionIterator delaying GC of the sub iterator and cleanup function (usually a closure referring to a larger collection).
if so, I'd open a separate JIRA+PR for this.